cli/http_util.rs

```diff
@@ -46,6 +50,200 @@ fn resolve_url_from_location(base_url: &Url, location: &str) -> Url {
 // Vec<(String, String)>
 pub type HeadersMap = HashMap<String, String>;
 
+type CacheControl = HashMap<String, Option<String>>;
+
+fn parse_cache_control(header: &str) -> CacheControl {
+  let mut cc = CacheControl::new();
+  let mut is_valid = true;
+  for part in header.split(',') {
+    if part.trim().is_empty() {
+      continue;
+    }
+    let mut kv = part.splitn(2, '=');
+    let k = kv.next().unwrap().trim();
+    if k.is_empty() {
+      continue;
+    }
+    let v = kv.next().map(str::trim);
+    match cc.entry(k.into()) {
+      Entry::Occupied(e) => {
+        if e.get().as_deref() != v {
+          is_valid = false;
+        }
+      }
+      Entry::Vacant(e) => {
+        e.insert(v.map(|v| v.trim_matches('"')).map(From::from));
+      }
+    }
+  }
+  if !is_valid {
+    cc.insert("must-revalidate".into(), None);
+  }
+  cc
+}
+
+/// A structure used to determine if a entity in the http cache can be used.
+///
+/// This is heavily influenced by
+/// https://github.com/kornelski/rusty-http-cache-semantics which is BSD
+/// 2-Clause Licensed and copyright Kornel Lesiński
+pub(crate) struct CacheSemantics {
+  cache_control: HashMap<String, Option<String>>,
+  cached: SystemTime,
+  headers: HashMap<String, String>,
+  now: SystemTime,
+}
+
+impl CacheSemantics {
+  pub fn new(
+    headers: HashMap<String, String>,
+    cached: SystemTime,
+    now: SystemTime,
+  ) -> Self {
+    let cache_control = headers
+      .get("cache-control")
+      .map(|v| parse_cache_control(v))
+      .unwrap_or_default();
+    Self {
+      cache_control,
+      cached,
+      headers,
+      now,
+    }
+  }
+
+  fn age(&self) -> Duration {
+    let mut age = self.age_header_value();
+
+    if let Ok(resident_time) = self.now.duration_since(self.cached) {
+      age += resident_time;
+    }
+
+    age
+  }
+
+  fn age_header_value(&self) -> Duration {
+    Duration::from_secs(
+      self
+        .headers
+        .get("age")
+        .and_then(|v| v.parse().ok())
+        .unwrap_or(0),
+    )
+  }
+
+  fn is_stale(&self) -> bool {
+    self.max_age() <= self.age()
+  }
+
+  fn max_age(&self) -> Duration {
+    if self.cache_control.contains_key("no-cache") {
```

Comment:
Where is this logic (and the one below) from? Is this from a specification somewhere?

Reply:
Based off of: https://docs.rs/http-cache-semantics/latest/src/http_cache_semantics/lib.rs.html#531-597

Reply:
Why don't we just use that crate?

Reply:
Because it deals only with request/response pairs which have quite different structures for things like headers.  It would require re-organizing the whole of the file fetcher/http cache to basically make it the `Cache` stuff and re-doing the whole of the way we persist stuff in the cache.